gourmet/gtk_extras/pango_html.py

```diff
@@ -0,0 +1,175 @@
+        # italics, bold, underlined, normal, and links (coloured and underlined)
+        soup = BeautifulSoup(tags, features="lxml")
+        tags = soup.find_all("tag")
+
+        tags_list = {}
+        for tag in tags:
+            opening_tags = ""
+            closing_tags = ""
+
+            # The tag may have a name, for named tags, or else an id
+            tag_name = tag.attrs.get('id')
+            tag_name = tag.attrs.get('name', tag_name)
+
+            attributes = [c for c in tag.contents if isinstance(c, Tag)]
+            for attribute in attributes:
+                vtype = attribute['type']
+                value = attribute['value']
+                name = attribute['name']
+
+                if vtype == "GdkColor":  # Convert colours to html
+                    if name in ['foreground-gdk', 'background-gdk']:
+                        opening, closing = self.tag2html[name]
+                        hex_color = self.pango_to_html_hex(value)
+                        opening = opening.format(hex_color)
+                    else:
+                        continue  # no idea!
+                else:
+                    opening, closing = self.tag2html[value]
+
+                opening_tags += opening
+                closing_tags = closing + closing_tags   # closing tags are FILO
+
+            tags_list[tag_name] = opening_tags, closing_tags
+
+            if opening_tags:
+                tags_list[tag_name] = opening_tags, closing_tags
+
+        self.tags = tags_list
+
+        # Create a single output string that will be sequentially appended to
+        # during feeding of text. It can then be returned once we've parse all
+        self.markup_text = ""
+        self.current_opening_tags = ""
+        self.current_closing_tags = ""
+        self.is_colored_and_underlined = False
+        super().feed(text)
+        return self.markup_text
+
+    def handle_starttag(self, tag: str, attrs: List[Tuple[str, str]]) -> None:
+        # The only tag in pango markup is "apply_tag". This could be ignored or
+        # made an assert, but we let our parser quietly handle nonsense.
+        if tag == "apply_tag":
+            attrs = dict(attrs)
+            tag_name = attrs.get('id')  # A tag may have a name, or else an id
+            tag_name = attrs.get('name', tag_name)
+            tags = self.tags.get(tag_name)
+
+            if tags is not None:
+                self.current_opening_tags, self.current_closing_tags = tags
+
+        if 'foreground' and '<u>' in self.current_opening_tags:
+            self.is_colored_and_underlined = True
+
+    def handle_data(self, data: str) -> None:
+        target = self.links.get(data)
+
+        if self.is_colored_and_underlined and target is not None:
+            # Replace the markup tags with a hyperlink target
+            data = f'<a href="{target}">{data}</a>'
+        else:
+            data = self.current_opening_tags + data + self.current_closing_tags
```

Comment:
This is incorrect.
There should be a queue of tags to close, and pop from that in `handle_endtag`.

An examples is:
`1<b>23<i>45</i>67</b>89`, which incorrectly gets serialised as `1<b>23</b><i>45</i>6789`

Reply:
```python
    pango_markup = b'GTKTEXTBUFFERCONTENTS-0001\x00\x00\x01i <text_view_markup>\n <tags>\n  <tag name="italic" priority="1">\n   <attr name="style" type="PangoStyle" value="PANGO_STYLE_ITALIC" />\n  </tag>\n  <tag name="bold" priority="0">\n   <attr name="weight" type="gint" value="700" />\n  </tag>\n </tags>\n<text>ddf<apply_tag name="bold">fd<apply_tag name="italic">df</apply_tag>fd</apply_tag>dff</text>\n</text_view_markup>\n'  # noqa
    expected = 'ddf<b>fd<i>df</i>fd</b>dff'

    ret = PangoToHtml().feed(pango_markup)
    assert ret == expected
```